lua/entities/gmod_wire_expression2/core/files.lua

```diff
@@ -22,226 +22,241 @@ E2Lib.registerConstant( "FILE_TIMEOUT", FILE_TIMEOUT )
+	end
 
 	return true
 end
 
-local function file_Download( ply, filename, data, append )
-	if !file_canDownload( ply ) or !IsValid( ply ) or !ply:IsPlayer() or !E2Lib.isValidFileWritePath(filename) then return false end
-	if string.len( data ) > (cv_max_transfer_size:GetInt() * 1024) then return false end
+local function file_canDownload(ply)
+	return #downloads[ply] < cv_transfer_max:GetInt()
+end
+
+local function file_Download(self, ply, filename, data, append)
+	if not file_canDownload(ply) then return self:throw("You've reached the file write limit!") end
+	if not E2Lib.isValidFileWritePath(filename) then return self:throw("Invalid file name/path: " .. filename) end
+	if #data > cv_max_transfer_size:GetInt() * 1024 then return self:throw("File is too large to send!") end
 
 	-- if we're trying to append an empty string then we don't need to queue up
 	-- a download to consider the operation completed
-	if append and string.len(data) == 0 then return true end
+	if append and #data == 0 then return 0 end
+
+	local queue = downloads[ply]
 
-	downloads[ply] = {
+	queue[#queue + 1] = {
 		name = filename,
 		data = data,
 		started = false,
 		downloading = true,
 		downloaded = false,
 		append = append
 	}
-end
-
 
-local function file_canList( ply )
-	local plist = lists[ply]
-	local pdel = (delays[ply] or {}).list
+	flushFileBuffer()
 
-	if (plist and plist.uploading) or
-		(pdel and (CurTime() - pdel) < cv_transfer_delay:GetInt()) then return false end
+	return #data
+end
 
-	return true
+local function file_canList(ply)
+	return #lists[ply] < cv_transfer_max:GetInt()
 end
 
 util.AddNetworkString("wire_expression2_request_list")
-local function file_List( ply, entity, dir )
-	if !file_canList( ply ) or !IsValid( ply ) or !ply:IsPlayer() then return false end
+local function file_List(self, ply, entity, dir)
+	if not file_canList(ply) then return self:throw("You've reached the file list download limit!") end
 
-	lists[ply] = {
+	local queue = lists[ply]
+	queue[#queue + 1] = {
 		dir = dir,
 		data = {},
 		uploading = true,
 		uploaded = false,
 		ent = entity
 	}
 	net.Start("wire_expression2_request_list")
-		net.WriteString( dir or "" )
+		net.WriteString(dir or "")
 	net.Send(ply)
 
-	delays[ply].list = CurTime()
+	return true
 end
 
 --- File loading ---
 
-__e2setcost( 20 )
+__e2setcost(100)
 
-e2function void fileLoad( string filename )
-	file_Upload( self.player, self.entity, filename )
+e2function number fileLoad(string filename)
+	return file_Upload(self, self.player, self.entity, filename) and 1 or 0
 end
 
-__e2setcost( 5 )
+__e2setcost(3)
 
+[nodiscard]
 e2function number fileCanLoad()
-	return file_canUpload( self.player ) and 1 or 0
+	return file_canUpload(self.player) and 1 or 0
 end
 
+[nodiscard]
+e2function number fileLoadQueued()
+	return #(downloads[self.player] or {})
+end
+
+__e2setcost(5)
+
+[deprecated]
 e2function number fileLoaded()
-	local pfile = uploads[self.player]
+	local pfile = uploads[self.player].last
```

Comment:
What happens if uploads[self.player] is nil? Do we need a nil check?

Reply:
Yeah, looks like these need nil checks.

Reply:
Nevermind, I just noticed the construct hook initializes the these tables for the player.